qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -616,8 +653,8 @@ def __init__(
         consts_expr = _sigma > 0
         valid_amp_conditions_expr = sym.Abs(_amp) <= 1.0
 
-        super().__init__(
-            pulse_type=self.__class__.__name__,
+        instance = SymbolicPulse(
+            pulse_type=cls.alias,
```

Comment:
Just checking: does `SymbolicPulse.pulse_type` have any mathematical meaning, or is it just like a "name" field?  I find it weird if it has mathematical meaning (the subclasses seem cleaner, in that case), but if it's just a name, then 👍.

Reply:
Good question. It could have mathematical meaning, because a symbolic pulse instance is tied to a particular symbolic expression, and we implicitly assumes unique mapping to `pulse_type`. This is not something user can change per instance. Note that this information is just kind of metadata, because waveform data can be generated without type information, i.e. `get_waveform` can generate sample data by itself. 

The motivation of this PR is to address round-trip serialization issue. The important point here is all pulse classes are not necessary defined in Qiskit. For example, in research code, end-user can define new pulse shape in parametric form, and they may choose not to expose it to Qiskit before publication. On the other hand, they may want to share the pulse with colleagues. 
In this situation, something weird happens with subclasses. If one defines `class MyPulse(SymbolicPulse)` and instantiates it, then serializes it through QPY and load at a later time. This comes back as a direct `SymbolicPulse` instance because Qiskit, particularly QPY, will never know the subclass, i.e. the `MyPulse` doesn't exist in `qiskit.pulse.library`. However, if everything is direct instance of `SymbolicPulse` and such type is defined by `.pulse_type: str` rather than class, we can serialize the data properly.

Perhaps you know better idea :)

Reply:
Note that an end-user is not necessary a single Qiskit user. For example, IBM backend may have several pulses, and experiment library should be able to handle these resource without modifying QPY loader for different libraries. Another option would be upgrade QPY to take Encoder/Decoder like JSON.

Reply:
Pulse names have a special meaning because the backend can support a set of pulses parametrically for which only the name and parameters need to be provided. For IBM backends, these names are snakecase versions of the old pulse classes (Gaussian <-> gaussian). Some of the awkwardness of `pulse_type` is that we are still kind of supporting the old classes by letting you create a gaussian pulse using `Gaussian` and have its repr show up as `Gaussian(...)` using the `pulse_type` (so invocation and repr stay the same, though the object is now a SymbolicPulse rather than a Gaussian). The `ParametricPulseShapes` enum translates the old class name to the name for the backend.

Ideally, one day the backend can accept a symbolic expression directly in the pulse data instead of just the name of a supported pulse and then we could send that expression and the names would not matter.

It would be nice if we could move the backend names into the SymbolicPulse data and drop the special-case ParametricPulseShapes translation.

Reply:
ugh, in current framework `pulse_type` is more than metadata. So these pulses should be subclasses as long as we can avoid round trip serialization issue. 

> It would be nice if we could move the backend names into the SymbolicPulse data and drop the special-case ParametricPulseShapes translation.

I'm bit against to this direction because this ties a particular backend to user program. Probably we can introduce an internal pulse class so that we can set backend name reported by backends in the transform logic.

Reply:
> I'm bit against to this direction because this ties a particular backend to user program. Probably we can introduce an internal pulse class so that we can set backend name reported by backends in the transform logic.

I don't understand the concern here. You don't like including something like `parametric_pulse_name: gaussian` in the Gaussian SymbolicPulse because `gaussian` implies it is the IBM backend? With these pulses defined in terra, couldn't any backend choose to implement them with the given names?

Reply:
Yes. Because the Qobj converter is standalone mechanism from assemble, and provider can give their pulse name through a custom converter. If we tie this to pulse class itself this is kind of enforcing the serialization rule to all provider and programs are no longer backend agnostic. I think Gaussian Square is kind of IBM convention, e.g. I more often see Gaussian flat top in literatures.